Add WebM support

zaps166 opened the pull request:
This PR adds WebM support for Godot: WebM (simplified MKV) container, VP8 and VP9 video codecs. It uses already existing Vorbis and Opus audio codecs.

YASM is required for all x86 platforms (without it, it fallback to very slow C code with a message in console).
All unsupported configurations should fallbacks to slow C code (e.g. JavaScript or MIPS CPU) with a message in console.

Used 3rd party libraries:
- libvpx - the VP8/VP9 library,
- libsimplewebm which already contains libwebm from libvpx-1.6.0/third_party - https://github.com/zaps166/libsimplewebm

Testing:
- **Not tested platforms: Haiku, BB10**.
- Compilation-only tests: Android x86/ARMv6, JavaScript, iPhone, WinRT(x86-x64 and ARM).
- Run-time basic tests: Linux/X11 32bit/64bit (GCC, Clang), Windows 7 VS2015 32bit/64bit, Windows 7 MinGW 32bit/64bit, OSX 32bit/64bit/fat, Android ARMv7+NEON.

TODO:
- bug fixes and compilation fixes.

zaps166 committed:
Add libvpx thirdparty library

Only necessary files

zaps166 committed:
libvpx: Necessary modifications

- convert ARM assembly (NEON) files,
- add rtcd for run-time CPU features detection,
- modify "system_state.h",
- "arm_cpudetect.c" fixes.

zaps166 commented:
@akien-mga 

Travis can't handle iPhone compilation (works properly locally on VirtualBox):

```
The log length has exceeded the limit of 4 MB (this usually means that the test suite is raising the same exception over and over).

The job has been terminated
```

I can't see any problems, so log limit on Travis should be changed - how to do this?

vnen commented on SConstruct:
```diff
@@ -339,6 +342,56 @@ if selected_platform in platform_list:
+		env['webm_cpu_arm'] = selected_platform == 'winrt' or selected_platform == 'iphone' or selected_platform == 'bb10' or (selected_platform == 'android' and env["android_arch"] != 'x86')
+		env['webm_cpu_x86'] = (env["bits"] == "32" or env["bits"] == "64") and (selected_platform == 'windows' or selected_platform == 'x11' or selected_platform == 'osx' or selected_platform == 'haiku' or (selected_platform == 'android' and env["android_arch"] == 'x86'))
+
+		if env['webm_cpu_x86'] and Execute("yasm --version"):
```
This `Execute` fails on Windows even with YASM in the PATH. Replacing it with `os.system` works.

zaps166 replied:
Ok, so I'll replace it, but weird, I've checked it on Windows (`yasm.exe` was in Godot directory).

vnen commented:
Working fine on Windows 10 64-bits,

zaps166 commented:
Ok, so I'll cleanup some commits and push fixed `SConstruct` for Windows later.

Btw. is it possible to kill this CI run: https://travis-ci.org/godotengine/godot/builds/160137721 (looped forever on iPhone compilation due to stdout/stderr log size :smile:) ?

vnen commented:
> is it possible to kill this CI run: https://travis-ci.org/godotengine/godot/builds/160137721

OK, I canceled it.

punto- commented:
I see a few allocations made with "new" and "new []", are these freed by
the library? If not we should use memnew, memdelete, memnew_arr and
memdelete_arr..

On 15 September 2016 at 18:19, George Marques <notifications@github.com>
wrote:

> is it possible to kill this CI run: https://travis-ci.org/
> godotengine/godot/builds/160137721
>
> OK, I canceled it.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/6490#issuecomment-247458589>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGVmPZzQOUrVF_3WTtT40iCK6Dl1qK4Oks5qqbZ0gaJpZM4J9QzG>
> .
>

zaps166 commented:
@punto- I'll check for memory leaks soon, but memory should be properly freed: https://github.com/zaps166/godot/blob/3716dd4f238f6731472e8d75baf203e537b0db01/drivers/webm/video_stream_webm.cpp#L144, https://github.com/zaps166/godot/blob/3716dd4f238f6731472e8d75baf203e537b0db01/drivers/webm/video_stream_webm.cpp#L359 I'll also check my `libsimplewebm` library. In 3rd party libraries it should not have memory leaks :)

Question about platforms:
- bb10 - is it still supported, how to build for it?
- javascript - doesn't work for me since Godot 2.1 (throws exception at startup in FF),
- haiku - I'm unable to test it,
- winrt - someone else should try to compile and run it.

Maybe WebM should be disabled for those platforms to prevent compilation errors? Also JavaScript should use browser built-in WebM support (is it possible via emscripten?) to achieve better performance than plain C code compiled into JS :) I assume that VP8/VP9 JS code will be unplayable on most CPUs.

zaps166 commented:
This PR is too big, I think is better to split this to three PRs:
- add `libyuv` and use it in theora,
- fix theora x86 SIMD compilation (in SCons),
- add WebM support which uses already existing `libyuv`.

zaps166 commented:
Updated!

Now this is based on #6616. WebM is **disabled** by default (use `webm=yes` to enable).

@vnen `Execute` is replaced by `os.system` in `SConstruct`.

@punto- I can't see memory leaks so far using address sanitizer.

punto- commented:
To clarify: I didn't mean there are memory leaks, I mean the engine has its
own memory allocators, we don't use "new" and "delete", we use memnew(),
memdelete(), etc. We should switch to those, and keep in mind that if the
library frees the objects that we allocate, it'll use "delete", which is
not compatible with memnew(), etc

On 26 September 2016 at 12:37, Błażej Szczygieł <notifications@github.com>
wrote:

> Updated!
>
> Now this is based on #6616
> <https://github.com/godotengine/godot/pull/6616>. WebM is *disabled* by
> default (use webm=yes to enable).
>
> @vnen <https://github.com/vnen> Execute is replaced by os.system in
> SConstruct.
>
> @punto- <https://github.com/punto-> I can't see memory leaks so far using
> address sanitizer.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/6490#issuecomment-249607091>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGVmPec-Vgc_r7PsIsDq4Z9n9uLRrY1pks5qt-bIgaJpZM4J9QzG>
> .
>

zaps166 commented:
@punto- 

`malloc()`, `realloc()`, `free()` are used for `video_frames` array. Replace it by `memrealloc`?

I'll replace my usage of `new` and `delete` in `video_stream_webm.cpp`.

> (...) and keep in mind that if the library frees the objects that we allocate, it'll use "delete", which is not compatible with memnew(), etc

Right, I'll check it after replacing `new` operator.

zaps166 commented:
@punto- C/C++ allocators are replaced by Godot's allocators whenever possible (not touching libraries).

Also fixed minor bugs.

akien-mga commented:
Note that #6616 should be merged before this one.

zaps166 commented:
Also I should do one more change in code and enable it in SCons for most platforms.

It should be reviewed by SCons expert :smile: Is it possible to add different CFLAGS only for several C sources?

punto- commented:
You can by copying the env into a new one and adding the flags to that.
Check out #4757 and how we do it for env_drivers, to avoid polluting the
main env with driver-specific flags.

On 3 October 2016 at 07:42, Błażej Szczygieł <notifications@github.com>
wrote:

> Also I should do one more change in code and enable it in SCons for most
> platforms.
>
> It should be reviewed by SCons expert 😄 Is it possible to add different
> CFLAGS only for several C sources?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/6490#issuecomment-251076707>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGVmPeL0jM2n-9jaYgQoiCe5qGSEoYR2ks5qwNwfgaJpZM4J9QzG>
> .
>

zaps166 commented:
@punto- I tried and it works :smile: Does Godot support non-SSE2 CPUs?

reduz commented:
no it doesn't

On Wed, Oct 5, 2016 at 9:30 AM, Błażej Szczygieł <notifications@github.com>
wrote:

> @punto- <https://github.com/punto-> I tried and it works 😄 Does Godot
> support non-SSE2 CPUs?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/6490#issuecomment-251660668>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z2wfwREDwOB7LwqWipag9yXQhWJRjks5qw5h0gaJpZM4J9QzG>
> .
>

zaps166 commented:
Updated (commits and first message with a description). If everything will be OK - this PR is ready :)
Every other improvements and bugfixes can be done later.

zaps166 commented:
There is BUG in SCons script, Android x86 64-bit is compiled by `i686-linux-android-gcc`, so it is not 64-bit, so it can't link with 64-bit ASM files (but it is not related to this PR)...

zaps166 commented:
CI failed on iPhone, but it compiles on my VirtualBox OS X 10.10, so I'll disable WebM it for Travis/iPhone.

punto- commented:
Disabled everything, or just the arm optimizations? it's ok to just leave
it using CPU where the optimizations don't compile, it cam be fixed later

On 5 October 2016 at 20:25, Błażej Szczygieł <notifications@github.com>
wrote:

> CI failed on iPhone, but it compiles on my VirtualBox OS X 10.10, so I'll
> disable WebM it for Travis/iPhone.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/6490#issuecomment-251828189>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGVmPUpSqRPhg6L4FpQhzqr3xQBPK5zbks5qxDHugaJpZM4J9QzG>
> .
>

zaps166 commented:
@punto- 
> Disabled everything, or just the arm optimizations?

Where and what disabled :) ?

For ARM CPUs, SIMD optimizations are always enabled and should compile (WinRT not tested), but something is wrong with the CI, because I've tested it on my OSX on VirtualBox and it compiles.

akien-mga commented:
@zaps166 Regarding the added libraries, I'm currently working on moving all third party code to `thirdparty/` and allow unbundling everything that can (i.e. building against system libraries on Linux when asked for): https://github.com/akien-mga/godot/tree/thirdparty

However it would not make much sense to rebase right now as it would conflict with my upcoming PR, but just so you know what is coming :)

zaps166 commented:
I'll rebase soon... :smile:

zaps166 committed:
Add libsimplewebm and libwebm thirdparty libraries

zaps166 committed:
Add WebM module

Use already existing libraries: libvorbis and libopus. Also use newly
added libraries: libvpx, libwebm, libsimplewebm.

zaps166 commented:
Rebased, now it uses old and slow `yuv2rgb.h` and **only** system `libvpx`.

akien-mga commented on thirdparty/README.md:
```diff
@@ -216,3 +216,8 @@ Files extracted from upstream source:
 
 - all .c and .h files apart from `gz*`
+
+## libsimplewebm
```
You might have noticed that this file is alphabetically ordered :P

akien-mga commented on thirdparty/libsimplewebm/SCsub:
```diff
@@ -0,0 +1,10 @@
+Import('env')
```
Probably not needed (for now) as `modules/webm/SCsub` builds it already?

zaps166 replied:
Yes, old code, sorry

zaps166 commented:
Built-in `libvpx` should be here, not in separate PR, right ?

akien-mga commented:
>Built-in libvpx should be here, not in separate PR, right ?

That would be best yes, so that in can build on platforms which don't provide it as a system library.

zaps166 commented:
@akien-mga 
Fully rebased (with libvpx), now SCons files are very simple thanks to your thrdparty libs separation :) Could you review especially SCons files and 3rd party licenses?

zaps166 commented:
Hmm, maybe armv6 assembly is not supported in new XCode... Also upstream libvpx has removed armv6 assembly.

zaps166 commented:
I'll remove armv6 assembly:
- simplify SCsub file,
- less code, less files,
- armv6 is old and also too slow for VP8 decoding,
- already removed upstream (https://github.com/webmproject/libvpx/commit/d55724fae9cb27e070add7952394fc0427ef2061).

zaps166 commented:
Almost ready, the only problem is with `yasm` detection on Windows.

zaps166 commented:
I think this PR is ready... :)

vnen commented:
Can't build for winrt x86_64. There's a wrong assumption that winrt always build for ARM, it's also possible to build for x86, since it runs on PCs too. There's already architecture detection in the platform code, so if need be it's easy to expose that to the environment so you can check in the module's code.

zaps166 commented:
@vnen
Ok! WindowsRT, Haiku and BB10 are not tested. How can I test it?

vnen commented:
Can't build for WinRT for ARM either. It tries to run some `ml` command to build the assembly files, but there are no such command. I don't have an ARM device though, so I can only test when I'm able to build for x86.

BTW, will this require some change in the docs for compiling (because of YASM)?

zaps166 commented:
> BTW, will this require some change in the docs for compiling (because of YASM)?

Hmm yes, `yasm` must be available for x86 (any x86 target, also Android) for optimized code.

---

Ah, `winrt` is `Universal Windows Platform` in docs :smile: (I don't know what is it, I've never used Windows Store or Windows Phone) I'll try to fix it :)

zaps166 commented:
@vnen 

Can't install Windows 10 SDK :disappointed:

> Windows 10 SDK (10.0.14393) : The installer failed. Błąd krytyczny podczas instalacji. Error code: -2147023293

Can I install Windows 10 SDK on Windows 7 SP1 ?

zaps166 commented:
I can run compilation (after installation error), but:
```

C:\godot>\Python27\Scripts\scons.bat p=winrt
scons: Reading SConscript files ...
Compiled program architecture will be an ARM executable. (forcing bits=32).
scons: done reading SConscript files.
scons: Building targets ...
cl /Foplatform\winrt\joystick_winrt.winrt.debug.32.arm.obj /c platform\winrt\joystick_winrt.cpp /TP /ZW /FS /nologo /Zi /DDEBUG_ENABLED /DDEBUG_MEMORY_ENABLED /FS /MP /GS /wd"4453" /wd"28204" /wd"4291
" /Zc:wchar_t /Gm- /fp:precise /D "_UNICODE" /D "UNICODE" /D "WINAPI_FAMILY=WINAPI_FAMILY_APP" /errorReport:prompt /WX- /Zc:forScope /Gd /EHsc /nologo /AI "C:\Program Files (x86)\Microsoft Visual Stud
io 14.0\VC\\vcpackages" /AI "C:\Program Files (x86)\Windows Kits\8.1\\References\CommonConfiguration\Neutral" /DWINRT_ENABLED /DWINDOWS_ENABLED /DTYPED_METHOD_BIND /DGLES2_ENABLED /DGL_GLEXT_PROTOTYPE
S /DEGL_EGLEXT_PROTOTYPES /DANGLE_ENABLED /FI "modules/freetype/winrtdef.h" -DFREETYPE_ENABLED -DDEBUG_MEMORY_ALLOC -DSCI_NAMESPACE -DENABLE_DEPRECATED /D __WRL_NO_DEFAULT_LIB__ /D WIN32 /FU "C:\Progr
am Files (x86)\Microsoft Visual Studio 14.0\VC\lib/store/references/platform.winmd" /AI "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\lib/store/references" /MDd -DGDSCRIPT_ENABLED -DMINIZIP_
ENABLED -DXML_ENABLED -DOPENSSL_ENABLED /Icore /Icore\math /Itools /Idrivers /I. /IC:\angle-ms-master\include /Iplatform\winrt /Idrivers\windows /Ithirdparty\zlib /Ithirdparty\freetype /Ithirdparty\fr
eetype\include /Ithirdparty\libpng /Ithirdparty\openssl
joystick_winrt.cpp
c:\godot\platform\winrt\joystick_winrt.h(58): error C3083: 'Gaming': the symbol to the left of a '::' must be a type
c:\godot\platform\winrt\joystick_winrt.h(58): error C3083: 'Input': the symbol to the left of a '::' must be a type
c:\godot\platform\winrt\joystick_winrt.h(58): error C2039: 'IGameController': is not a member of 'Windows'
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE\vccorlib.h(2893): note: see declaration of 'Windows'
c:\godot\platform\winrt\joystick_winrt.h(58): error C2143: syntax error: missing ';' before '^'
c:\godot\platform\winrt\joystick_winrt.h(58): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
c:\godot\platform\winrt\joystick_winrt.h(58): error C2238: unexpected token(s) preceding ';'
c:\godot\platform\winrt\joystick_winrt.h(75): error C3083: 'Gaming': the symbol to the left of a '::' must be a type
c:\godot\platform\winrt\joystick_winrt.h(75): error C3083: 'Input': the symbol to the left of a '::' must be a type
c:\godot\platform\winrt\joystick_winrt.h(75): error C2039: 'Gamepad': is not a member of 'Windows'
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE\vccorlib.h(2893): note: see declaration of 'Windows'
c:\godot\platform\winrt\joystick_winrt.h(75): error C2061: syntax error: identifier 'Gamepad'
c:\godot\platform\winrt\joystick_winrt.h(76): error C3083: 'Gaming': the symbol to the left of a '::' must be a type
c:\godot\platform\winrt\joystick_winrt.h(76): error C3083: 'Input': the symbol to the left of a '::' must be a type
c:\godot\platform\winrt\joystick_winrt.h(76): error C2039: 'Gamepad': is not a member of 'Windows'
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE\vccorlib.h(2893): note: see declaration of 'Windows'
c:\godot\platform\winrt\joystick_winrt.h(76): error C2061: syntax error: identifier 'Gamepad'
platform\winrt\joystick_winrt.cpp(32): error C3083: 'Gaming': the symbol to the left of a '::' must be a type
platform\winrt\joystick_winrt.cpp(32): error C2039: 'Input': is not a member of 'Windows'
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE\vccorlib.h(2893): note: see declaration of 'Windows'
platform\winrt\joystick_winrt.cpp(32): error C2871: 'Input': a namespace with this name does not exist
platform\winrt\joystick_winrt.cpp(37): error C2653: 'Gamepad': is not a class or namespace name
platform\winrt\joystick_winrt.cpp(37): error C2065: 'GamepadAdded': undeclared identifier
platform\winrt\joystick_winrt.cpp(38): error C2065: 'Gamepad': undeclared identifier
platform\winrt\joystick_winrt.cpp(38): error C2059: syntax error: '>'
platform\winrt\joystick_winrt.cpp(38): error C2976: 'Windows::Foundation::EventHandler': too few generic arguments
platform\winrt\joystick_winrt.cpp(38): error C2955: 'Windows::Foundation::EventHandler': use of class generic requires generic argument list
platform\winrt\joystick_winrt.cpp(38): error C2664: 'Windows::Foundation::EventHandler::EventHandler(Platform::Object ^,Platform::IntPtr)': cannot convert argument 2 from 'void (__cdecl JoystickWinrt:
:* )(Platform::Object ^)' to 'Platform::IntPtr'
platform\winrt\joystick_winrt.cpp(38): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
platform\winrt\joystick_winrt.cpp(39): error C2653: 'Gamepad': is not a class or namespace name
platform\winrt\joystick_winrt.cpp(39): error C2065: 'GamepadRemoved': undeclared identifier
platform\winrt\joystick_winrt.cpp(40): error C2065: 'Gamepad': undeclared identifier
platform\winrt\joystick_winrt.cpp(40): error C2059: syntax error: '>'
platform\winrt\joystick_winrt.cpp(40): error C2976: 'Windows::Foundation::EventHandler': too few generic arguments
platform\winrt\joystick_winrt.cpp(40): error C2955: 'Windows::Foundation::EventHandler': use of class generic requires generic argument list
platform\winrt\joystick_winrt.cpp(40): error C2664: 'Windows::Foundation::EventHandler::EventHandler(Platform::Object ^,Platform::IntPtr)': cannot convert argument 2 from 'void (__cdecl JoystickWinrt:
:* )(Platform::Object ^)' to 'Platform::IntPtr'
platform\winrt\joystick_winrt.cpp(40): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
platform\winrt\joystick_winrt.cpp(53): error C2065: 'GamepadReading': undeclared identifier
platform\winrt\joystick_winrt.cpp(53): error C2146: syntax error: missing ';' before identifier 'reading'
platform\winrt\joystick_winrt.cpp(53): error C2065: 'reading': undeclared identifier
platform\winrt\joystick_winrt.cpp(53): error C2065: 'Gamepad': undeclared identifier
platform\winrt\joystick_winrt.cpp(53): error C2059: syntax error: ')'
platform\winrt\joystick_winrt.cpp(55): error C2653: 'GamepadButtons': is not a class or namespace name
platform\winrt\joystick_winrt.cpp(55): error C2065: 'Menu': undeclared identifier
platform\winrt\joystick_winrt.cpp(58): error C2065: 'reading': undeclared identifier
platform\winrt\joystick_winrt.cpp(58): error C2228: left of '.Buttons' must have class/struct/union
platform\winrt\joystick_winrt.cpp(58): note: type is 'unknown-type'
platform\winrt\joystick_winrt.cpp(58): error C2660: 'InputDefault::joy_button': function does not take 3 arguments
platform\winrt\joystick_winrt.cpp(62): error C2065: 'reading': undeclared identifier
platform\winrt\joystick_winrt.cpp(62): error C2228: left of '.LeftThumbstickX' must have class/struct/union
platform\winrt\joystick_winrt.cpp(62): note: type is 'unknown-type'
platform\winrt\joystick_winrt.cpp(62): error C2660: 'InputDefault::joy_axis': function does not take 3 arguments
platform\winrt\joystick_winrt.cpp(63): error C2065: 'reading': undeclared identifier
platform\winrt\joystick_winrt.cpp(63): error C2228: left of '.LeftThumbstickY' must have class/struct/union
platform\winrt\joystick_winrt.cpp(63): note: type is 'unknown-type'
platform\winrt\joystick_winrt.cpp(63): error C2660: 'InputDefault::joy_axis': function does not take 3 arguments
platform\winrt\joystick_winrt.cpp(64): error C2065: 'reading': undeclared identifier
platform\winrt\joystick_winrt.cpp(64): error C2228: left of '.RightThumbstickX' must have class/struct/union
platform\winrt\joystick_winrt.cpp(64): note: type is 'unknown-type'
platform\winrt\joystick_winrt.cpp(64): error C2660: 'InputDefault::joy_axis': function does not take 3 arguments
platform\winrt\joystick_winrt.cpp(65): error C2065: 'reading': undeclared identifier
platform\winrt\joystick_winrt.cpp(65): error C2228: left of '.RightThumbstickY' must have class/struct/union
platform\winrt\joystick_winrt.cpp(65): note: type is 'unknown-type'
platform\winrt\joystick_winrt.cpp(65): error C2660: 'InputDefault::joy_axis': function does not take 3 arguments
platform\winrt\joystick_winrt.cpp(66): error C2065: 'reading': undeclared identifier
platform\winrt\joystick_winrt.cpp(66): error C2228: left of '.LeftTrigger' must have class/struct/union
platform\winrt\joystick_winrt.cpp(66): note: type is 'unknown-type'
platform\winrt\joystick_winrt.cpp(66): error C2660: 'InputDefault::joy_axis': function does not take 3 arguments
platform\winrt\joystick_winrt.cpp(67): error C2065: 'reading': undeclared identifier
platform\winrt\joystick_winrt.cpp(67): error C2228: left of '.RightTrigger' must have class/struct/union
platform\winrt\joystick_winrt.cpp(67): note: type is 'unknown-type'
platform\winrt\joystick_winrt.cpp(67): error C2660: 'InputDefault::joy_axis': function does not take 3 arguments
platform\winrt\joystick_winrt.cpp(90): error C3083: 'Gaming': the symbol to the left of a '::' must be a type
platform\winrt\joystick_winrt.cpp(90): error C3083: 'Input': the symbol to the left of a '::' must be a type
platform\winrt\joystick_winrt.cpp(90): error C2039: 'Gamepad': is not a member of 'Windows'
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE\vccorlib.h(2893): note: see declaration of 'Windows'
platform\winrt\joystick_winrt.cpp(90): error C2061: syntax error: identifier 'Gamepad'
platform\winrt\joystick_winrt.cpp(105): error C2039: 'controller_reference': is not a member of 'JoystickWinrt::ControllerDevice'
c:\godot\platform\winrt\joystick_winrt.h(56): note: see declaration of 'JoystickWinrt::ControllerDevice'
platform\winrt\joystick_winrt.cpp(105): error C2065: 'value': undeclared identifier
platform\winrt\joystick_winrt.cpp(112): error C3083: 'Gaming': the symbol to the left of a '::' must be a type
platform\winrt\joystick_winrt.cpp(112): error C3083: 'Input': the symbol to the left of a '::' must be a type
platform\winrt\joystick_winrt.cpp(112): error C2039: 'Gamepad': is not a member of 'Windows'
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE\vccorlib.h(2893): note: see declaration of 'Windows'
platform\winrt\joystick_winrt.cpp(112): error C2061: syntax error: identifier 'Gamepad'
platform\winrt\joystick_winrt.cpp(118): error C2039: 'controller_reference': is not a member of 'JoystickWinrt::ControllerDevice'
c:\godot\platform\winrt\joystick_winrt.h(56): note: see declaration of 'JoystickWinrt::ControllerDevice'
platform\winrt\joystick_winrt.cpp(118): error C2065: 'value': undeclared identifier
scons: *** [platform\winrt\joystick_winrt.winrt.debug.32.arm.obj] Error 2
scons: building terminated because of errors.

C:\godot>
```
I have no idea how to compile for WinRT...

vnen commented:
@zaps166 there's probably something missing. You should be able to install the Windows SDK through the Visual Studio installer.

zaps166 commented:
@vnen 
![spectacle l17319](https://cloud.githubusercontent.com/assets/6239090/19576898/2fd11fe6-9715-11e6-9e90-8ffc5fb04885.png)

vnen commented:
@zaps166 might be related to Windows 7... Does it have updates installed?

zaps166 commented:
@vnen I've got always updated disabled, this is Windows on VirtualBox (I use Linux). I'll try with clean Windows 10 installation, it will take some time... :)

zaps166 commented:
@vnen I can install this on Windows 10, I don't know why it fails on Windows 7, maybe disabled services or missing updates or something else is wrong :smile:

Now it should compile under WinRT x86 and ARM - please test! :)

Btw. what exactly is "bits=fat" and how can I use it?

vnen commented:
Ok, working for me on Windows and UWP/WinRT.

IIRC `bits=fat` is for iOS and OSX. It bundles the 32- and 64-bits binaries in a single file.

zaps166 commented:
@vnen Ok, so `fat` build for OSX is not optimized now, I'll fix it soon :)

zaps166 commented:
`p=osx bits=fat` should now work properly

punto- commented:
winrt has an "arch" parameter doesn't it?

On 20 October 2016 at 14:41, George Marques <notifications@github.com>
wrote:

> Can't build for winrt x86_64. There's a wrong assumption that winrt always
> build for ARM, it's also possible to build for x86, since it runs on PCs
> too. There's already architecture detection in the platform code, so if
> need be it's easy to expose that to the environment so you can check in the
> module's code.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/6490#issuecomment-255175772>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGVmPW5uxAi4F1ODR4KsRoH1wdV8nvHbks5q16evgaJpZM4J9QzG>
> .
>

punto- commented:
how does the fat build work when 2 architectures require different
parameters to build?

On 21 October 2016 at 20:39, Błażej Szczygieł <notifications@github.com>
wrote:

> @vnen <https://github.com/vnen> Ok, so fat build for OSX is not optimized
> now, I'll fix it soon :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/6490#issuecomment-255489910>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGVmPXh-b-t19z-uhEWTkpmnNNt6WKh4ks5q2U0-gaJpZM4J9QzG>
> .
>

zaps166 commented:
@punto- 
> how does the fat build work when 2 architectures require different parameters to build?

`yasm` will use the same parameters every time (external script `yasm_osx_fat.py`), C/C++ code should use different parameters. Please give me an example how can I use different parameters for architectures in `fat` mode, then I'll test it.

> winrt has an "arch" parameter doesn't it?

Yes, `arch` is in `winrt/detect.py` it can be `x86`, `x64` or `arm`. It is not exported, but I'm using it in this way: `if 'arm' in env["PROGSUFFIX"]:` to detect an ARM architecture for WinRT, because `arch` occurs in `PROGSUFFIX`.

vnen commented:
@punto- WinRT has no `arch` parameter. It detects the architecture based on the environment (which is set by an argument to `vcvarsall.bat`). BTW, this batch file sets an environment variable `PLATFORM=ARM` which can also be used to detect.

akien-mga commented:
Let's merge, as I'm about to apply pep8 to all the buildsystem, and I think @zaps166 is tired of rebasing to take into account my refactorings :p

If some stuff still needs improvement, please open an issue about it so that it isn't forgotten.

akien-mga commented:
@zaps166 Next time you work on the webm support, please add in `thirdparty/README.md` the version information for `libwebm` :)

zaps166 commented:
@akien-mga Ok! I forgot to check the libwebm version (IIRC I've took it from libvpx 1.6.0 3rdparty sources) :)

Corruptinator commented:
I'm just curious, would this merged pr allow the engine to playback any online video, if done by extensive GDScripting? A.K.A. Youtube, etc? or is just an alternative video format instead of OGG/OGV?

Calinou commented:
> I'm just curious, would this merged PR allow the engine to playback any online video, if done by extensive GDScripting? A.K.A. Youtube, etc? or is just an alternative video format instead of OGG/OGV?

It allows playing back VP8/VP9 videos along with Vorbis/Opus audio, using the CPU only. It works successfully so far in my tests (but uses a lot of CPU in 1080p, prefer using 720p for now, especially on mobile devices) including audio.

Downloading videos from YouTube or most other video sharing sites is a complicated subject, and I'm afraid it's not realistically possible to stream them in Godot.

Corruptinator commented:
Thanks for the reply! I just wanted to know if it's possible but I guess its just a long stretch from something like that happening anytime in the future.

Keyaku commented:
@Calinou How's performance in comparison to what we had before?

In any case, thank you @zaps166, @akien-mga and @punto- for this one!

Calinou commented:
> How's performance in comparison to what we had before?

It is not as good as Theora performance, because Theora is easy to decode for the CPU, but it is inefficient, both in file size and image quality.

Keyaku commented:
@Calinou Are you sure? I was expecting a perofrmance increase from what read.

Bummer.

zaps166 commented:
CPU performance can be improved by using `libyuv` (#6616), or better, by using GLSL for YUV->RGB conversion. Unfortunately, `libvpx` has a little worse performance than FFmpeg VP8/VP9 decoder :(

The demuxer and decoder from this PR is able to demux/decode online (also youtube) streams, but it needs a modification - it must read from the network stream instead of resource. Also not all youtube movies are easy to play, most (not all due to protection) 360p VP8 with Vorbis audio are simple to extract URL and play. Higher video resolutions with VP9 are in separate streams for audio and video.

reduz commented:
I was thinking of properly supporting yuv textures in 3.0

On Oct 31, 2016 17:30, "Błażej Szczygieł" <notifications@github.com> wrote:

> CPU performance can be improved by using libyuv (#6616
> <https://github.com/godotengine/godot/pull/6616>), or better, by using
> GLSL for YUV->RGB conversion. Unfortunately, libvpx has a little worse
> performance than FFmpeg VP8/VP9 decoder :(
>
> The demuxer and decoder from this PR is able to demux/decode online (also
> youtube) streams, but it needs a modification - it must read from the
> network stream instead of resource. Also not all youtube movies are easy to
> play, most (not all due to protection) 360p VP8 with Vorbis audio are
> simple to extract URL and play. Higher video resolutions with VP9 are in
> separate streams for audio and video.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/6490#issuecomment-257411633>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AF-Z296-NmnE7CZWV0jPGC0ALrSVx-cDks5q5k_7gaJpZM4J9QzG>
> .
>

reduz commented:
Problem is that you need to allocate both the size of the yuv and the
decompressed texture.. not so terrible probably though

On Oct 31, 2016 17:33, "Juan Linietsky" <reduzio@gmail.com> wrote:

> I was thinking of properly supporting yuv textures in 3.0
>
> On Oct 31, 2016 17:30, "Błażej Szczygieł" <notifications@github.com>
> wrote:
>
>> CPU performance can be improved by using libyuv (#6616
>> <https://github.com/godotengine/godot/pull/6616>), or better, by using
>> GLSL for YUV->RGB conversion. Unfortunately, libvpx has a little worse
>> performance than FFmpeg VP8/VP9 decoder :(
>>
>> The demuxer and decoder from this PR is able to demux/decode online (also
>> youtube) streams, but it needs a modification - it must read from the
>> network stream instead of resource. Also not all youtube movies are easy to
>> play, most (not all due to protection) 360p VP8 with Vorbis audio are
>> simple to extract URL and play. Higher video resolutions with VP9 are in
>> separate streams for audio and video.
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/godotengine/godot/pull/6490#issuecomment-257411633>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AF-Z296-NmnE7CZWV0jPGC0ALrSVx-cDks5q5k_7gaJpZM4J9QzG>
>> .
>>
>

vnen commented:
@zaps166 have you ever built this for iPhone x86? Had a report about compilation error on the Facebook group that was solved by disabling this module.

zaps166 commented:
@vnen No, I thought this is impossible :D I'll fix it. Which command line arguments can compile for iPhone x86 ?

vnen commented:
@zaps166 add `ios_sim=yes` to the command. This builds for the simulator, which runs in x86 architecture.

zaps166 commented:
@vnen Ok, fixed :) #7123
